stdlib/SparseArrays/src/sparsematrix.jl

```diff
@@ -373,12 +373,12 @@ SparseMatrixCSC(M::AbstractMatrix{Tv}) where {Tv} = SparseMatrixCSC{Tv,Int}(M)
 SparseMatrixCSC{Tv}(M::AbstractMatrix{Tv}) where {Tv} = SparseMatrixCSC{Tv,Int}(M)
 function SparseMatrixCSC{Tv,Ti}(M::AbstractMatrix) where {Tv,Ti}
     @assert !has_offset_axes(M)
-    I = findall(x -> x != 0, M)
-    eltypeTiI = Ti[i[1] for i in I]
-    eltypeTiJ = Ti[i[2] for i in I]
-    eltypeTvV = Tv[M[i] for i in I]
-    return sparse_IJ_sorted!(eltypeTiI, eltypeTiJ, eltypeTvV, size(M)...)
+    C = findall(v->v != 0, M)
+    @inbounds I = convert(Vector{Ti},LinearIndices(M)[C])
+    @inbounds V = convert(Vector{Tv},M[C])
```

Comment:
You can do both `I` and `V` with a single loop and no temporary arrays with:
```jl
I = Ti[]
V = Tv[]
Mi = LinearIndices(M)
for (i, v) in pairs(M)
    if !iszero(v)
        push!(I, Mi[i])
        push!(V, v)
    end
end
```

Reply:
I've had tried this initially, and I recall it to be worse performance wise. I thought it was because the predictive reallocation had to be done twice (for I and V) during pushing. But no, I just tried yours and it is much faster. I was probably doing something else wrongly.

Reply:
Spoke too soon. Redoing the test with `Transpose` above, I get (copied master for easier comparison)

master
-------------------------------------------------------------
|    n\p        | 1.0         |   0.1     | 0.01          |  0.001       | 0.0001
|-----------|------------|-----------|---------------|---------------|-------------
| n=10     | 1.481 μs | 517.677 ns | 379.346 ns | 350.789 ns | 353.089 ns
| n=100    | 121.878 μs | 37.005 μs | 20.916 μs | 18.598 μs | 18.476 μs
| n=1000   | 20.618 ms | 6.356 ms | 3.295 ms | 2.892 ms | 2.936 ms
----------------------------------------------------------------------------------

push! version
-------------------------------------------------------------
|    n\p        | 1.0         |   0.1     | 0.01          |  0.001       | 0.0001
|-----------|------------|-----------|---------------|---------------|-------------
| n=10     | 1.434 μs | 528.342 ns | 416.040 ns | 359.900 ns | 359.172 ns
| n=100    | 131.639 μs | 56.422 μs | 33.416 μs | 29.433 μs | 28.957 μs
| n=1000   | 19.793 ms | 6.810 ms | 4.015 ms | 3.612 ms | 3.510 ms
----------------------------------------------------------------------------------

Reply:
The test in which the `push!` version was faster was using the bare functions with `M<:Matrix` (without wrapped stuff)... I don't understand the difference.

Reply:
Is the problem the `pairs` function for `Transpose`?  What about:
```
I = Ti[]
V = Tv[]
i = 0
for v in M
    i += 1
    if !iszero(v)
        push!(I, i)
        push!(V, v)
    end
end

Reply:
(Of course, looping by columns in a `Transpose` array is hard on the cache, but there's not much to be done about that here.   However, by the same token I'm not sure that the performance in the `Transpose` case is a major concern for this function.)

Reply:
> Is the problem the `pairs` function for `Transpose`? What about:

Yes! This is slightly faster than mine in the `Transpose` case. Thanks!

Reply:
> (Of course, looping by columns in a `Transpose` array is hard on the cache, but there's not much to be done about that here. However, by the same token I'm not sure that the performance in the `Transpose` case is a major concern for this function.)

Exactly, do you have a better benchmark in mind?

Reply:
Just a regular `Matrix` would be the most relevant case, I would think?

Reply:
`Matrix` is catched by the `SparseMatrixCSC{Tv,Ti}(M::StridedMatrix)` constructor, which is faster.